### ISSUE TITLE
Classification pipeline suggestion

### ISSUE THREAD

**Issue by rbbby** [welfare-state-analytics/riksdagen-corpus#182]
### Summary
- We need a structured pipeline for classifications that will be easy to maintain
- Classifiers should be trained on OCR:d data to the extent possible in order to avoid serial dependencies
- We should aim to limit the number of classifiers to reduce costs and to avoid model dependencies and conflicts.

### Current pipeline
Essentially we want to build classifiers using active learning to solve many tasks in the corpus. This task is not entirely trivial as exemplified by the current pipeline looking something like:

> OCR:rd data --> note/seg classification --> intro regex detection (with splitting of intros) --> intro bert detection (with splitting of intros) --> merging of intros

Predicting intro merges was even done before the second splitting of intros, but these predictions are applied later in the pipeline (would have to rerun the prediction model to fix this conflict). As it is right now, classifiers are trained on data from earlier stages in the pipeline (and from different versions of the corpus), so they are essentially trained conditional on other models predictions (and regex).

### Problem
Sequentiality and unclear dependencies makes the pipeline fragile. We want to update parts earlier in the pipeline (add classifiers, retrain classifiers, fix bugs, regex, etc.). The effect of doing so is that the data changes and models suddenly predicts on data it has not been trained on. While this may not seem like a huge problem, it gets out of hand quickly once we add more models (and is even more of a mess considering conflicts as the one mentioned earlier). This is also a corpus quality issue, as if we have a classifier with a test set accuracy of 95% in a late stage of the pipeline and update something in an earlier stage, suddenly we have new data and the classifiers accuracy becomes unknown.

### Solution
Essentially, our ad-hoc approach has worked fine so far but has some issues already and will spiral out of control becoming increasingly fragile, have more uncertain/incorrect estimates of the quality, and be more difficult and time consuming to work with. I argue that we should avoid having sequential dependencies for the classifiers as much as possible, and train almost all of them on the OCR:d data. This also brings multiple additional benefits:

- We can make use of KB:s new efficient annotation tool
- We can utilise metadata like OCR bounding box coordinates and images (we currently lack a mapping back to the original data and I argue that we should not try to map back in #179, but rather update the OCR --> corpus pipeline)
- We can add more classes in a good way
- We can have 1 or at least limit the number of datasets for the classification tasks (which decreases the cost of AL significantly)
- This limits the number of models that has to make costly predictions

Our training data atm is also partially incorrectly annotated, it is from different releases, and it lacks mapping back to both the OCR:d data and corpus(!!!). The data even contains observations that do not exist anymore due to pipeline conflicts.

### Which classifiers?
Currently we have a note/seg classifier that needs to be updated. We want classifiers for debate topics and potentially some other new classes. The introduction classifier could be improved using metadata (type-1 errors during lists of speakers).
I suggest we try out a multilabel classifier to do all of these things. It would take text, images and metadata as input and output which classes a given text block belongs to. We can start out without images but they are easy to add later on. I suggest a model like this since:

- It probably works and would eliminate the need for some other models
- We only need 1 dataset (and to perform AL for 1 model)
- It is easier to update in the future and avoids sequentiality
- It handles ambiguous classes caused by segmentation issues
- It costs very little in the long run since we will have really good training data to try out other approaches with if it fails at some task

### Schedule
In order to deploy such a model, we need the alto --> corpus pipeline working (discussed in #179). @ninpnin I assume that we have it in place but just need some minor updates?
We should probably check if we want to re:ocr some time period before annotating data from that period.
Starting the process of training a classifier can be done in parallell and I already have most of this pipeline done from another project.

This would be a relatively big project with a couple releases, but I think this is the direction we want to go, a lot of this has to be done anyways, and this is an efficient way of progressing.
